tools/topology/topology2/imx8-wm8960.conf

```diff
@@ -0,0 +1,128 @@
+##
+## Example topology for i.MX8MP board with 'CODEC' codec
+##
+## (source) host.N.playback -> buffer.N.1 -> volume.N.1 -> buffer.N.2 -> SAI.SAI_DAI_INDEX(sink endpoint)
+##
+
+# Include paths
+<searchdir:include>
+<searchdir:include/common>
+<searchdir:include/components>
+<searchdir:include/controls>
+<searchdir:include/dais>
+<searchdir:include/pipelines>
+
+# Include class definitions
+<vendor-token.conf>
+<hw_config_simple.conf>
+<tokens.conf>
+<volume-playback.conf>
+<dai.conf>
+<data.conf>
+<pcm.conf>
+<pcm_caps.conf>
+<fe_dai.conf>
+<sai.conf>
+<manifest.conf>
+<route.conf>
+
+# DAI definition
+Object.Dai {
+	SAI.0 {
+		id					0
+		dai_index			$SAI_DAI_INDEX
+		direction			"playback"
+		name				$STREAM_CODEC_NAME
+		default_hw_conf_id	0
+
+		Object.Base.hw_config.0 {
+			id				0
+			name			"wm8960"
+			mclk_freq		12288000
+			bclk_freq		3072000
+			tdm_slot_width	32
+			format			"I2S"
+			bclk			"codec_provider"
+			fsync			"codec_provider"
+			fsync_freq		48000
+			tdm_slots		2
+			tx_slots		3
+			rx_slots		3
+		}
+
+		Object.Widget.dai.1 {
+			direction					"playback"
+			index						1
+			type						dai_in
+			stream_name					$STREAM_CODEC_NAME
+			period_sink_count			0
+			period_source_count			2
+			format						s32le
+			num_input_audio_formats		1
+			num_output_audio_formats	1
+			num_input_pins				1
+		}
+	}
+}
+
+#
+# Pipeline definitions
+#
+# (source) host.N.playback -> buffer.N.1 -> volume.N.1 -> buffer.N.2 (sink endpoint)
+#
+Object.Pipeline {
+	volume-playback.1 {
+		index	1
+
+		Object.Widget.pipeline.1 {
+			stream_name		'dai.SAI.0.playback'
+		}
+
+		Object.Widget.host.playback {
+			stream_name		'Passthrough Playback'
+		}
+
+		Object.Widget.pga.1 {
+			Object.Control.mixer.1 {
+				name	'Playback Volume'
+			}
+		}
+
+		Object.Widget.buffer.1 {
```

Comment:
Should I use `buffer` here?
I don't see buffer module in any toml and I'm getting `failed updating IPC struct for buffer.1.2` in Linux.

So, maybe I should use `host-copier` and `dai-copier`?

If I use these new component, I won't need buffers, right?
If I understood correctly the buffers are encapsulated in the host/dai-copier component.

Reply:
@iuliana-prodan with IPC4, the intermediate buffers are no longer needed in topology. Those will be added by the firmware automatically when connecting modules. So in your simple case, you could just have a pipeline that looks like host-copier->gain->dai-copier.

Reply:
Thank you @ranj063.
I'll try the host-copier->gain->dai-copier.